src/i_oalsound.c

```diff
@@ -0,0 +1,631 @@
+//
+// Copyright(C) 2023 Roman Fomin
+// Copyright(C) 2023 ceski
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//      System interface for OpenAL sound.
+//
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "doomstat.h"
+#include "i_oalsound.h"
+#include "r_data.h"
+
+#define OAL_ROLLOFF_FACTOR 1
+#define OAL_SPEED_OF_SOUND 343.3f
+// 128 map units per 3 meters (https://doomwiki.org/wiki/Map_unit).
+#define OAL_MAP_UNITS_PER_METER (128.0f / 3.0f)
+#define OAL_SOURCE_RADIUS 32.0f
+#define OAL_DEFAULT_PITCH 1.0f
+#define OAL_NUM_ATTRIBS 5
+
+#define VOL_TO_GAIN(x) ((ALfloat)(x) / 127)
+
+// C doesn't allow casting between function and non-function pointer types, so
+// with C99 we need to use a union to reinterpret the pointer type. Pre-C99
+// still needs to use a normal cast and live with the warning (C++ is fine with
+// a regular reinterpret_cast).
+#if __STDC_VERSION__ >= 199901L
+#define FUNCTION_CAST(T, ptr) (union{void *p; T f;}){ptr}.f
+#else
+#define FUNCTION_CAST(T, ptr) (T)(ptr)
+#endif
+
+int snd_resampler;
+boolean snd_hrtf;
+int snd_absorption;
+int snd_doppler;
+
+boolean oal_use_doppler;
+
+static const char *oal_resamplers[] = {
+    "Nearest", "Linear", "Cubic"
+};
+
+typedef struct oal_system_s
+{
+    ALCdevice *device;
+    ALCcontext *context;
+    ALuint *sources;
+    boolean *active;
+    ALuint *buffers;
```

Comment:
Since we store the OpenAL buffer IDs in `oal_system_t`, perhaps we no longer need to store them in `sfxinfo_t`?

Reply:
The interaction between `channelinfo` and `sfxinfo_t` in `i_source.c` is tricky so I'll have to look at that again. The buffers in `i_oalsound.c` is there just as a convenient way to free them all at once with `alDeleteBuffers()` when calling `I_OAL_ShutdownSound()`.

Reply:
> The interaction between `channelinfo` and `sfxinfo_t` in `i_source.c` is tricky

Yes, I think we can simplify or even remove most of it.

Reply:
Whatever you guys say. Would this be subject to this PR, or is this a different issue?

Reply:
I think we can postpone this to another PR.

Reply:
Agreed.